File: app/src/main/java/com/ldt/musicr/ui/page/librarypage/song/SongChildTab.java
package com.ldt.musicr.ui.page.librarypage.song;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.Group;
import androidx.core.view.OnApplyWindowInsetsListener;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.ldt.musicr.App;
import com.ldt.musicr.R;
import com.ldt.musicr.contract.AbsMediaAdapter;
import com.ldt.musicr.loader.medialoader.SongLoader;
import com.ldt.musicr.model.Song;
import com.ldt.musicr.ui.page.MusicServiceFragment;
import com.ldt.musicr.ui.bottomsheet.SortOrderBottomSheet;
import com.ldt.musicr.util.InterpolatorUtil;
import com.ldt.musicr.util.Tool;
import com.ldt.musicr.util.Util;
import com.simplecityapps.recyclerview_fastscroll.views.FastScrollRecyclerView;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class SongChildTab extends MusicServiceFragment implements SortOrderBottomSheet.SortOrderChangedListener, PreviewRandomPlayAdapter.FirstItemCallBack {
    public static final String TAG = "SongChildTab";

    @BindView(R.id.recycler_view)
    RecyclerView mRecyclerView;

//    @BindView(R.id.preview_shuffle_list)
//    RecyclerView mPreviewRecyclerView;

    @BindView(R.id.refresh)
    ImageView mRefresh;

    @BindView(R.id.image)
    ImageView mImage;

    @BindView(R.id.title)
    TextView mTitle;

    @BindView(R.id.description)
    TextView mArtist;

    @BindView(R.id.random_group)
    Group mRandomGroup;

    private int mCurrentSortOrder = 0;

    private void initSortOrder() {
        mCurrentSortOrder = App.getInstance().getPreferencesUtility().getSongChildSortOrder();
    }
//    @BindView(R.id.top_background) View mTopBackground;
//    @BindView(R.id.bottom_background) View mBottomBackground;
//    @BindView(R.id.random_header) View mRandomHeader;
//    @BindView(R.id.shuffle_button) View ShuffleButton;


    @OnClick({R.id.preview_random_panel})
    void shuffle() {
        mAdapter.shuffle();
    }

    private final SongChildAdapter mAdapter = new SongChildAdapter();
//    PreviewRandomPlayAdapter mPreviewAdapter;

    @OnClick(R.id.refresh)
    void refresh() {
        mRefresh.animate().rotationBy(360).setInterpolator(InterpolatorUtil.getInterpolator(6)).setDuration(650);
        mRefresh.postDelayed(mAdapter::randomize, 300);
    }


    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.screen_songs_tab, container, false);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mAdapter.init(requireContext());
        mAdapter.setName(TAG);
        mAdapter.setCallBack(this);
        mAdapter.setSortOrderChangedListener(this);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ButterKnife.bind(this, view);
        initSortOrder();
        ViewCompat.setOnApplyWindowInsetsListener(mRecyclerView, new OnApplyWindowInsetsListener() {
            @Override
            public WindowInsetsCompat onApplyWindowInsets(View v, WindowInsetsCompat insets) {
                v.setPadding(insets.getSystemWindowInsetLeft(),
                        0,
                        insets.getSystemWindowInsetRight(),
                        (int) (insets.getSystemWindowInsetBottom() + v.getResources().getDimension(R.dimen.bottom_back_stack_spacing)));
                return ViewCompat.onApplyWindowInsets(v, insets);
            }
        });
        mRecyclerView.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false));
        mRecyclerView.setAdapter(mAdapter);

        refreshData();
    }

    @Override
    public void onDestroyView() {
        mAdapter.destroy();
        super.onDestroyView();
    }

    private void refreshData() {
    /*    if(getContext() != null)
        SongLoader.doSomething(getContext());
*/
        ArrayList<Song> songs = SongLoader.getAllSongs(getActivity(), SortOrderBottomSheet.mSortOrderCodes[mCurrentSortOrder]);
        mAdapter.setData(songs);
        showOrHidePreview(!songs.isEmpty());

    }

    private void showOrHidePreview(boolean show) {
        int v = show ? View.VISIBLE : View.GONE;

        mRandomGroup.setVisibility(v);
    }

    @Override
    public void onFirstItemCreated(Song song) {
        mTitle.setText(song.title);
        mArtist.setText(song.artistName);

        Glide.with(this)
                .load(Util.getAlbumArtUri(song.albumId))
                .placeholder(R.drawable.music_style)
                .error(R.drawable.music_empty)
                .into(mImage);

    }

    @Override
    public void onPlayingMetaChanged() {
        if (mAdapter != null)
            mAdapter.notifyOnMediaStateChanged(AbsMediaAdapter.PLAY_STATE_CHANGED);
    }

    @Override
    public void onPaletteChanged() {
        if (mRecyclerView instanceof FastScrollRecyclerView) {
            FastScrollRecyclerView recyclerView = ((FastScrollRecyclerView) mRecyclerView);
            recyclerView.setPopupBgColor(Tool.getHeavyColor());
            recyclerView.setThumbColor(Tool.getHeavyColor());
        }
        mAdapter.notifyOnMediaStateChanged(AbsMediaAdapter.PALETTE_CHANGED);
        super.onPaletteChanged();
    }

    @Override
    public void onPlayStateChanged() {
        if (mAdapter != null)
            mAdapter.notifyOnMediaStateChanged(AbsMediaAdapter.PLAY_STATE_CHANGED);
    }

    @Override
    public void onMediaStoreChanged() {
        ArrayList<Song> songs = SongLoader.getAllSongs(getActivity(), SortOrderBottomSheet.mSortOrderCodes[mCurrentSortOrder]);
        mAdapter.setData(songs);
        showOrHidePreview(!songs.isEmpty());
    }

    @Override
    public int getSavedOrder() {
        return mCurrentSortOrder;
    }

    @Override
    public void onOrderChanged(int newType, String name) {
        if (mCurrentSortOrder != newType) {
            mCurrentSortOrder = newType;
            App.getInstance().getPreferencesUtility().setSongChildSortOrder(mCurrentSortOrder);
            refreshData();
        }
    }
}

File: app/src/main/java/com/ldt/musicr/ui/maintab/library/song/SongChildTab.kt
package com.ldt.musicr.ui.maintab.library.song

import android.content.Context
import com.ldt.musicr.ui.bottomsheet.SortOrderBottomSheet.SortOrderChangedListener
import com.ldt.musicr.ui.maintab.library.song.PreviewRandomPlayAdapter.FirstItemCallBack
import com.ldt.musicr.R
import androidx.recyclerview.widget.RecyclerView
import android.widget.TextView
import com.ldt.musicr.App
import com.ldt.musicr.util.InterpolatorUtil
import android.os.Bundle
import android.view.View
import android.widget.ImageView
import androidx.constraintlayout.widget.Group
import androidx.core.view.ViewCompat
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import com.ldt.musicr.model.Song
import com.ldt.musicr.loader.medialoader.SongLoader
import com.ldt.musicr.ui.bottomsheet.SortOrderBottomSheet
import com.bumptech.glide.Glide
import com.ldt.musicr.interactors.AppExecutors
import com.ldt.musicr.interactors.postOnUiThread
import com.ldt.musicr.model.dataitem.DataItem
import com.ldt.musicr.notification.Action
import com.ldt.musicr.notification.ActionResponder
import com.ldt.musicr.notification.EventKey
import com.ldt.musicr.ui.maintab.library.adapter.MediaAdapter
import com.ldt.musicr.util.Tool
import com.ldt.musicr.util.Util
import com.simplecityapps.recyclerview_fastscroll.views.FastScrollRecyclerView
import com.zalo.gitlabmobile.notification.MessageEvent
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe

class SongChildTab : Fragment(R.layout.screen_songs_tab), SortOrderChangedListener, FirstItemCallBack, ActionResponder {
    private lateinit var recyclerView: RecyclerView

    private lateinit var refreshView: ImageView
    private lateinit var imageView: ImageView
    private lateinit var titleTextView: TextView
    private lateinit var artistTextView: TextView
    private lateinit var randomGroup: Group
    private var currentSortOrder = 0

    private fun initSortOrder() {
        currentSortOrder = App.getInstance().preferencesUtility.songChildSortOrder
    }

    fun shuffle() {
        //adapter.shuffle()
    }

    private val adapter: MediaAdapter = MediaAdapter()

    private fun refresh() {
        refreshView.animate().rotationBy(360f).setInterpolator(InterpolatorUtil.getInterpolator(6)).duration = 650
        refreshView.postDelayed({ /*adapter.randomize()*/ }, 300)
    }

    private fun initView(view: View) {
        recyclerView = view.findViewById(R.id.recycler_view)
        refreshView = view.findViewById(R.id.refresh)
        imageView = view.findViewById(R.id.image)
        titleTextView = view.findViewById(R.id.title)
        artistTextView = view.findViewById(R.id.description)
        randomGroup = view.findViewById(R.id.random_group)

        view.findViewById<View>(R.id.preview_random_panel).setOnClickListener { shuffle() }
        view.findViewById<View>(R.id.refresh).setOnClickListener { refresh() }

        recyclerView.layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        recyclerView.adapter = adapter
        ViewCompat.setOnApplyWindowInsetsListener(recyclerView) { v, insets ->
            v.setPadding(
                insets.systemWindowInsetLeft,
                0,
                insets.systemWindowInsetRight,
                (insets.systemWindowInsetBottom + v.resources.getDimension(R.dimen.bottom_back_stack_spacing)).toInt()
            )
            ViewCompat.onApplyWindowInsets(v, insets)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        adapter.actionResponder = this
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initView(view)
        initSortOrder()
        refreshData()
    }

    private fun refreshData() {
        AppExecutors.io().execute {
            val songs = SongLoader.getAllSongs(App.getInstance().applicationContext, SortOrderBottomSheet.mSortOrderCodes[currentSortOrder])
            val result = mutableListOf<DataItem>()

            // Header
            result.add(DataItem.SortingTile)

            // Data
            songs.forEachIndexed { index, song -> result.add(DataItem.SongItem(song, index)) }

            //Finish
            postOnUiThread {
                if(isAdded && !isRemoving) {
                    showOrHidePreview(songs.isNotEmpty())
                    adapter.submitList(result)
                }
            }

        }
    }

    private fun showOrHidePreview(show: Boolean) {
        val v = if (show) View.VISIBLE else View.GONE
        randomGroup.visibility = v
    }

    override fun onFirstItemCreated(song: Song) {
        titleTextView.text = song.title
        artistTextView.text = song.artistName
        Glide.with(this)
            .load(Util.getAlbumArtUri(song.albumId.toLong()))
            .placeholder(R.drawable.ic_music_style)
            .error(R.drawable.music_empty)
            .into(imageView)
    }

    override fun getSavedOrder(): Int {
        return currentSortOrder
    }

    override fun onOrderChanged(newType: Int, name: String) {
        if (currentSortOrder != newType) {
            currentSortOrder = newType
            App.getInstance().preferencesUtility.songChildSortOrder = currentSortOrder
            refreshData()
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        EventBus.getDefault().register(this)
    }

    override fun onDetach() {
        super.onDetach()
        EventBus.getDefault().unregister(this)
    }

    @Subscribe
    fun onReceivedEvent(messageEvent: MessageEvent) {
        when(messageEvent.key) {
            EventKey.OnMediaStoreChanged -> {
                refreshData()
            }
            EventKey.OnPaletteChanged -> {
                if (recyclerView is FastScrollRecyclerView) {
                    val recyclerView = recyclerView as FastScrollRecyclerView
                    recyclerView.setPopupBgColor(Tool.getHeavyColor())
                    recyclerView.setThumbColor(Tool.getHeavyColor())
                }
                //adapter.notifyOnMediaStateChanged(AbsMediaAdapter.PALETTE_CHANGED)
            }
            EventKey.OnPlayStateChanged -> {
                //adapter.notifyOnMediaStateChanged(AbsMediaAdapter.PLAY_STATE_CHANGED)
            }
            EventKey.OnPlayingMetaChanged -> {
                //adapter.notifyOnMediaStateChanged(AbsMediaAdapter.PLAY_STATE_CHANGED)
            }
            EventKey.OnQueueChanged -> { }
            EventKey.OnRepeatModeChanged -> { }
            EventKey.OnServiceConnected -> { }
            EventKey.OnServiceDisconnected -> { }
            EventKey.OnShuffleModeChanged -> { }
            EventKey.Other -> { }
        }
    }

    override fun invoke(action: Action<Any>): Action<Any>? {
        return null
    }

    companion object {
        const val TAG = "SongChildTab"
    }
}